File: web/src/services/gb28181.ts
import { request } from './api';
import { buildQuery } from '../utils/query';

// ============= GB28181 设备管理 =============

export interface GB28181Device {
  id: string;
  device_code: string;
  register_address: string;
  register_port: number;
  sip_id: string;
  sip_domain: string;
  last_register_at: string | null;
  last_heartbeat_at: string | null;
  last_catalog_at: string | null;
  heartbeat_interval: number;
  status: string;
  channel_count: number;
  manufacturer: string;
  model: string;
  firmware: string;
  created_at: string;
  updated_at: string;
}

export interface GB28181DeviceUpdateRequest {
  sip_id?: string;
  sip_domain?: string;
  sip_password?: string;
  heartbeat_interval?: number;
}

export interface GB28181DeviceCreateRequest {
  device_code: string;
  sip_id?: string;
  sip_domain: string;
  sip_password?: string;
  heartbeat_interval?: number;
}

export interface CatalogTaskResponse {
  task_id: string;
}

export interface CatalogTaskStatusResponse {
  task_id: string;
  status: 'pending' | 'completed' | 'failed';
  channel_count: number;
  error?: string;
}

export interface GB28181DeviceListParams {
  page?: number;
  page_size?: number;
  keyword?: string;
  status?: string;
}

type PageResult<T> = { list: T[]; total: number; page: number; page_size: number };

/** GB28181 设备列表 */
export function listGB28181Devices(params: GB28181DeviceListParams) {
  return request<PageResult<GB28181Device>>(`/gb28181/devices${buildQuery({ page: params.page, page_size: params.page_size, keyword: params.keyword, status: params.status })}`);
}

/** 创建 GB28181 设备 */
export function createGB28181Device(data: GB28181DeviceCreateRequest) {
  return request<GB28181Device>('/gb28181/devices', {
    method: 'POST',
    body: JSON.stringify(data),
  });
}

/** GB28181 设备详情 */
export function getGB28181Device(id: string) {
  return request<GB28181Device>(`/gb28181/devices/${id}`);
}

/** 更新 GB28181 设备 */
export function updateGB28181Device(id: string, data: GB28181DeviceUpdateRequest) {
  return request<GB28181Device>(`/gb28181/devices/${id}`, {
    method: 'PUT',
    body: JSON.stringify(data),
  });
}

/** 删除 GB28181 设备 */
export function deleteGB28181Device(id: string) {
  return request<null>(`/gb28181/devices/${id}`, { method: 'DELETE' });
}

/** 批量删除 GB28181 设备 */
export function batchDeleteGB28181Devices(ids: string[]) {
  return request<{ total: number; success: number; failed: number }>('/gb28181/devices/batch-delete', {
    method: 'POST',
    body: JSON.stringify({ ids }),
  });
}

/** 触发目录查询 */
export function triggerCatalog(id: string) {
  return request<CatalogTaskResponse>(`/gb28181/devices/${id}/catalog`, { method: 'POST' });
}

/** 查询目录任务状态 */
export function getCatalogTaskStatus(taskId: string) {
  return request<CatalogTaskStatusResponse>(`/gb28181/catalog-tasks/${taskId}`);
}

/** 设备通道列表 */
export interface GB28181Channel {
  id: string;
  gb28181_device_id: string;
  device_name: string;
  status: string;
  manufacturer: string;
  model: string;
  channel_id: string;
  parental_id: string;
}

export function getGB28181Channels(id: string) {
  return request<GB28181Channel[]>(`/gb28181/devices/${id}/channels`);
}

// ============= GB28181 媒体 =============

export interface PlayResponse {
  url: string;
  protocol: string;
  stream_id: string;
}

/** 启动实时预览 */
export function startGB28181Live(deviceId: string) {
  return request<PlayResponse>('/media/gb28181/live/start', {
    method: 'POST',
    body: JSON.stringify({ device_id: deviceId }),
  });
}

/** 停止实时预览 */
export function stopGB28181Live(deviceId: string, streamId: string) {
  return request<null>('/media/gb28181/live/stop', {
    method: 'POST',
    body: JSON.stringify({ device_id: deviceId, stream_id: streamId }),
  });
}

/** 启动回放 */
export function startGB28181Playback(deviceId: string, startTime: string, endTime: string) {
  return request<PlayResponse>('/media/gb28181/playback/start', {
    method: 'POST',
    body: JSON.stringify({ device_id: deviceId, start_time: startTime, end_time: endTime }),
  });
}

/** 回放控制 */
export function controlGB28181Playback(streamId: string, action: string, speed?: number, stamp?: number) {
  return request<null>('/media/gb28181/playback/control', {
    method: 'POST',
    body: JSON.stringify({ stream_id: streamId, action, speed, stamp }),
  });
}

/** 停止回放 */
export function stopGB28181Playback(deviceId: string, streamId: string) {
  return request<null>('/media/gb28181/playback/stop', {
    method: 'POST',
    body: JSON.stringify({ device_id: deviceId, stream_id: streamId }),
  });
}

// ============= GB28181 配置 =============

// ============= GB28181 NVR（统一 Device 表） =============

/** GB28181 NVR 设备列表（从 Device 表查询） */
export function listGB28181NVRs(params: GB28181DeviceListParams) {
  return request<PageResult<GB28181Device>>(`/gb28181/nvrs${buildQuery({ page: params.page, page_size: params.page_size, keyword: params.keyword, status: params.status })}`);
}

/** 查询 NVR 下的所有通道 */
export function getGB28181NVRChannels(nvrId: string) {
  return request<GB28181Channel[]>(`/gb28181/nvrs/${nvrId}/channels`);
}

export function getGB28181Config() {
  return request<Record<string, string | number>>('/system/gb28181/config');
}

File: app/internal/handler/gb28181.go
package handler

import (
	"context"
	"encoding/json"
	"time"

	"github.com/gin-gonic/gin"
	"github.com/google/uuid"

	"github.com/niko-admin/niko-admin/internal/dto"
	"github.com/niko-admin/niko-admin/internal/model"
	"github.com/niko-admin/niko-admin/internal/pkg/cache"
	"github.com/niko-admin/niko-admin/internal/pkg/response"
	"github.com/niko-admin/niko-admin/internal/pkg/ws"
)

type gb28181Service interface {
	CreateGB28181Device(ctx context.Context, req dto.GB28181DeviceCreateRequest) (*model.GB28181Device, error)
	ListGB28181Devices(ctx context.Context, req dto.GB28181DeviceListRequest) ([]model.GB28181Device, int64, error)
	GetGB28181DeviceByID(ctx context.Context, id string) (*model.GB28181Device, error)
	UpdateGB28181Device(ctx context.Context, id string, req dto.GB28181DeviceUpdateRequest) (*model.GB28181Device, error)
	DeleteGB28181Device(ctx context.Context, id string) error
	BatchDeleteGB28181Devices(ctx context.Context, ids []string) dto.BatchResult
	QueryCatalog(ctx context.Context, deviceCode string) error
	GetGB28181DeviceChannels(ctx context.Context, id string) ([]model.Device, error)
	ListNVRs(ctx context.Context, keyword, status string, page, pageSize int) ([]dto.GB28181DeviceResponse, int64, error)
	GetNVRChannels(ctx context.Context, nvrID string, page, pageSize int) ([]dto.GB28181DeviceResponse, int64, error)
}

type GB28181Handler struct {
	sipSvc           gb28181Service
	cache            cache.Cache
	hub              *ws.Hub
}

func NewGB28181Handler(sipSvc gb28181Service, cache cache.Cache, hub *ws.Hub) *GB28181Handler {
	return &GB28181Handler{
		sipSvc:           sipSvc,
		cache:            cache,
		hub:              hub,
	}
}

// Create 创建 GB28181 设备
// @Summary      创建 GB28181 设备
// @Tags         GB28181管理
// @Accept       json
// @Produce      json
// @Param        body  body  dto.GB28181DeviceCreateRequest  true  "设备信息"
// @Success      200   {object}  dto.Response{data=dto.GB28181DeviceResponse}
// @Router       /gb28181/devices [post]
// @Security     BearerAuth
func (h *GB28181Handler) Create(c *gin.Context) {
	var req dto.GB28181DeviceCreateRequest
	if err := c.ShouldBindJSON(&req); err != nil {
		attachError(c, badRequestError(c, err))
		return
	}

	item, err := h.sipSvc.CreateGB28181Device(c.Request.Context(), req)
	if err != nil {
		attachError(c, err)
		return
	}

	response.OK(c, gb28181DeviceToResponse(item))
}

// List GB28181 设备列表
// @Summary      GB28181 设备列表
// @Tags         GB28181管理
// @Produce      json
// @Param        page        query   int     false  "页码"
// @Param        page_size   query   int     false  "每页数量"
// @Param        keyword     query   string  false  "关键词搜索"
// @Param        status      query   string  false  "状态筛选"
// @Success      200  {object}  dto.Response{data=dto.PageData{list=[]dto.GB28181DeviceResponse}}
// @Router       /gb28181/devices [get]
// @Security     BearerAuth
func (h *GB28181Handler) List(c *gin.Context) {
	var req dto.GB28181DeviceListRequest
	if err := c.ShouldBindQuery(&req); err != nil {
		attachError(c, badRequestError(c, err))
		return
	}

	items, total, err := h.sipSvc.ListGB28181Devices(c.Request.Context(), req)
	if err != nil {
		attachError(c, err)
		return
	}

	var list []dto.GB28181DeviceResponse
	for _, item := range items {
		list = append(list, gb28181DeviceToResponse(&item))
	}

	response.Page(c, list, total, req.GetPage(), req.GetPageSize())
}

// GetByID GB28181 设备详情
// @Summary      GB28181 设备详情
// @Tags         GB28181管理
// @Produce      json
// @Param        id  path  string  true  "设备 ID"
// @Success      200  {object}  dto.Response{data=dto.GB28181DeviceResponse}
// @Router       /gb28181/devices/{id} [get]
// @Security     BearerAuth
func (h *GB28181Handler) GetByID(c *gin.Context) {
	id := c.Param("id")
	item, err := h.sipSvc.GetGB28181DeviceByID(c.Request.Context(), id)
	if err != nil {
		attachError(c, err)
		return
	}

	response.OK(c, gb28181DeviceToResponse(item))
}

// Update 更新 GB28181 设备
// @Summary      更新 GB28181 设备
// @Tags         GB28181管理
// @Accept       json
// @Produce      json
// @Param        id    path   string                   true  "设备 ID"
// @Param        body  body   dto.GB28181DeviceUpdateRequest  true  "更新信息"
// @Success      200   {object}  dto.Response{data=dto.GB28181DeviceResponse}
// @Router       /gb28181/devices/{id} [put]
// @Security     BearerAuth
func (h *GB28181Handler) Update(c *gin.Context) {
	id := c.Param("id")
	var req dto.GB28181DeviceUpdateRequest
	if err := c.ShouldBindJSON(&req); err != nil {
		attachError(c, badRequestError(c, err))
		return
	}

	item, err := h.sipSvc.UpdateGB28181Device(c.Request.Context(), id, req)
	if err != nil {
		attachError(c, err)
		return
	}

	response.OK(c, gb28181DeviceToResponse(item))
}

// Delete 删除 GB28181 设备
// @Summary      删除 GB28181 设备
// @Tags         GB28181管理
// @Produce      json
// @Param        id  path  string  true  "设备 ID"
// @Success      200  {object}  dto.Response
// @Router       /gb28181/devices/{id} [delete]
// @Security     BearerAuth
func (h *GB28181Handler) Delete(c *gin.Context) {
	id := c.Param("id")
	if err := h.sipSvc.DeleteGB28181Device(c.Request.Context(), id); err != nil {
		attachError(c, err)
		return
	}
	response.OK(c, nil)
}

// BatchDelete 批量删除 GB28181 设备
// @Summary      批量删除 GB28181 设备
// @Tags         GB28181管理
// @Accept       json
// @Produce      json
// @Param        body  body  dto.BatchIDsRequest  true  "设备 ID 列表"
// @Success      200   {object}  dto.Response{data=dto.BatchResult}
// @Router       /gb28181/devices/batch-delete [post]
// @Security     BearerAuth
func (h *GB28181Handler) BatchDelete(c *gin.Context) {
	var req dto.BatchIDsRequest
	if err := c.ShouldBindJSON(&req); err != nil {
		attachError(c, badRequestError(c, err))
		return
	}
	response.OK(c, h.sipSvc.BatchDeleteGB28181Devices(c.Request.Context(), req.IDs))
}

// TriggerCatalog 触发目录查询
// @Summary      触发目录查询
// @Tags         GB28181管理
// @Produce      json
// @Param        id  path  string  true  "设备 ID"
// @Success      200  {object}  dto.Response{data=dto.CatalogTaskResponse}
// @Router       /gb28181/devices/{id}/catalog [post]
// @Security     BearerAuth
func (h *GB28181Handler) TriggerCatalog(c *gin.Context) {
	id := c.Param("id")
	item, err := h.sipSvc.GetGB28181DeviceByID(c.Request.Context(), id)
	if err != nil {
		attachError(c, err)
		return
	}

	taskID := uuid.New().String()
	cacheKey := "catalog_task:" + taskID
	deviceTaskKey := "catalog_task:device:" + item.DeviceCode

	status := dto.CatalogTaskStatusResponse{
		TaskID: taskID,
		Status: "pending",
	}
	statusData, _ := json.Marshal(status)
	_ = h.cache.Set(context.Background(), cacheKey, statusData, 2*time.Minute)
	_ = h.cache.Set(context.Background(), deviceTaskKey, []byte(taskID), 2*time.Minute)

	if err := h.sipSvc.QueryCatalog(c.Request.Context(), item.DeviceCode); err != nil {
		attachError(c, err)
		return
	}

	response.OK(c, dto.CatalogTaskResponse{TaskID: taskID})
}

// GetChannels 通道列表接口
// @Summary      通道列表接口
// @Tags         GB28181管理
// @Produce      json
// @Param        id  path  string  true  "设备 ID"
// @Success      200  {object}  dto.Response{data=[]dto.DeviceResponse}
// @Router       /gb28181/devices/{id}/channels [get]
// @Security     BearerAuth
func (h *GB28181Handler) GetChannels(c *gin.Context) {
	id := c.Param("id")
	devices, err := h.sipSvc.GetGB28181DeviceChannels(c.Request.Context(), id)
	if err != nil {
		attachError(c, err)
		return
	}

	var list []dto.DeviceResponse
	for _, item := range devices {
		list = append(list, dto.DeviceResponse{
			ID:               item.ID,
			DeviceName:       item.DeviceName,
			AccessType:       item.AccessType,
			GB28181DeviceID:  item.GB28181DeviceID,
			GB28181ChannelID: item.GB28181ChannelID,
			Status:           item.Status,
			Enabled:          item.Enabled,
			Manufacturer:     item.Manufacturer,
			Model:            item.Model,
			FirmwareVersion:  item.FirmwareVersion,
			CreatedAt:        item.CreatedAt.Format(time.RFC3339),
			UpdatedAt:        item.UpdatedAt.Format(time.RFC3339),
		})
	}
	response.OK(c, list)
}

// GetCatalogTaskStatus 状态查询接口
// @Summary      目录查询任务状态
// @Tags         GB28181管理
// @Produce      json
// @Param        task_id  path  string  true  "任务 ID"
// @Success      200  {object}  dto.Response{data=dto.CatalogTaskStatusResponse}
// @Router       /gb28181/catalog-tasks/{task_id} [get]
// @Security     BearerAuth
func (h *GB28181Handler) GetCatalogTaskStatus(c *gin.Context) {
	taskID := c.Param("task_id")
	data, err := h.cache.Get(context.Background(), "catalog_task:"+taskID)
	if err != nil {
		// If not found, it might have expired or never existed
		attachError(c, err)
		return
	}

	var status dto.CatalogTaskStatusResponse
	_ = json.Unmarshal(data, &status)

	response.OK(c, status)
}

// ListNVRs 列出 GB28181 NVR 设备（从统一 Device 表查询）
// @Summary      GB28181 NVR 设备列表（统一模型）
// @Tags         GB28181管理
// @Produce      json
// @Param        page        query   int     false  "页码"
// @Param        page_size   query   int     false  "每页数量"
// @Param        keyword     query   string  false  "关键词搜索"
// @Param        status      query   string  false  "状态筛选"
// @Success      200  {object}  dto.Response{data=dto.PageData}
// @Router       /gb28181/nvrs [get]
// @Security     BearerAuth
func (h *GB28181Handler) ListNVRs(c *gin.Context) {
	var req dto.GB28181DeviceListRequest
	if err := c.ShouldBindQuery(&req); err != nil {
		attachError(c, badRequestError(c, err))
		return
	}
	
	list, total, err := h.sipSvc.ListNVRs(c.Request.Context(), req.Keyword, req.Status, req.GetPage(), req.GetPageSize())
	if err != nil {
		attachError(c, err)
		return
	}
	
	response.Page(c, list, total, req.GetPage(), req.GetPageSize())
}

// GetNVRChannels 查询指定 NVR 下的所有通道
// @Summary      查询 NVR 下的 GB28181 通道
// @Tags         GB28181管理
// @Produce      json
// @Param        id  path  string  true  "NVR 设备 ID"
// @Param        page        query   int     false  "页码"
// @Param        page_size   query   int     false  "每页数量"
// @Success      200  {object}  dto.Response{data=dto.PageData}
// @Router       /gb28181/nvrs/{id}/channels [get]
// @Security     BearerAuth
func (h *GB28181Handler) GetNVRChannels(c *gin.Context) {
	nvrID := c.Param("id")

	var req dto.PageRequest
	if err := c.ShouldBindQuery(&req); err != nil {
		attachError(c, badRequestError(c, err))
		return
	}

	list, total, err := h.sipSvc.GetNVRChannels(c.Request.Context(), nvrID, req.GetPage(), req.GetPageSize())
	if err != nil {
		attachError(c, err)
		return
	}

	response.Page(c, list, total, req.GetPage(), req.GetPageSize())
}

func gb28181DeviceToResponse(item *model.GB28181Device) dto.GB28181DeviceResponse {
	return dto.GB28181DeviceResponse{
		ID:                item.ID,
		DeviceCode:        item.DeviceCode,
		RegisterAddress:   item.RegisterAddress,
		RegisterPort:      item.RegisterPort,
		SipID:             item.SipID,
		SipDomain:         item.SipDomain,
		LastRegisterAt:    item.LastRegisterAt,
		LastHeartbeatAt:   item.LastHeartbeatAt,
		LastCatalogAt:     item.LastCatalogAt,
		HeartbeatInterval: item.HeartbeatInterval,
		Status:            item.Status,
		ChannelCount:      item.ChannelCount,
		Manufacturer:      item.Manufacturer,
		Model:             item.Model,
		Firmware:          item.Firmware,
		CreatedAt:         item.CreatedAt,
		UpdatedAt:         item.UpdatedAt,
	}
}
